res/controllers/DJ TechTools-MIDI Fighter Twister-scripts.js

```diff
@@ -54,6 +54,157 @@ var MidiFighterTwister;
+                        0xf07800: 74,  // Orange
+                        0xfd4a4a: 85,  // Red (was being mapped to orange, perceptually closer to red)
+                        0xb90908: 85,  // Red
+                        0xff00ff: 100, // Fuscia
+                        0xa400c0: 108, // Purple
+                    }),
+                    type: components.Button.prototype.types.toggle,
+                    outValueScale: function(_value) {
+                        return this.colorMapper.getValueForNearestColor(engine.getParameter(this.group, this.colorKey));
+                    },
+                    shift: function() {
+                        this.inKey = "volume_set_default";
+                        this.outKey = "volume_set_default";
+                        this.type = components.Button.prototype.types.push;
+                        this.disconnect();
+                        this.connect();
+                        this.trigger();
+                    },
+                    unshift: function() {
+                        this.inKey = this.key;
+                        this.outKey = this.key;
+                        this.type = components.Button.prototype.types.toggle;
+                        this.disconnect();
+                        this.connect();
+                        this.trigger();
+                    },
+                    output: function(value, _group, _control) {
+                        if (value) {
+                            // Dim the LED
+                            midi.sendShortMsg(0xB2, this.midi[1], 35);
+                        } else {
+                            // Brighten the LED
+                            midi.sendShortMsg(0xB2, this.midi[1], 47);
+                        }
+                    },
+                    outputColor: function(colorCode) {
+                        const nearestColorValue = this.colorMapper.getValueForNearestColor(colorCode);
+                        this.send(nearestColorValue);
+                    },
+                    connect: function() {
+                        components.Button.prototype.connect.call(this);
+                        if (undefined !== this.group && this.colorKey !== undefined) {
+                            this.connections[1] = engine.makeConnection(this.group, this.colorKey, this.outputColor.bind(this));
+                        }
+                    },
+                });
+                this.effect[i] = new components.Encoder({
+                    group: `[QuickEffectRack1_[Channel${deckNums[0]}_Stem${i + 1}]]`,
+                    midi: [0xB0, this.midiModifier(0x21 + (4 * i))],
+                    key: "super1",
+                    inValueScale: function(value) {
+                        if (this.inKey === "loaded_chain_preset") {
+                            const val = (value/this.max) * engine.getParameter(this.group, "num_chain_presets");
+                            return val;
```

Comment:
consider manually rounding/truncating here. Qt might complain when you're trying to load chain preset 3.14 (it'll still work, but its gonna be unhappy about converting a non-integer float to an int).